### ISSUE TITLE
Support uneven DDP inputs

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#33148]
We have seen multiple users hit the problem of different DDP instance having different number of input batches. [[e.g.](https://discuss.pytorch.org/t/best-practice-for-uneven-dataset-sizes-with-distributeddataparallel/67308)] As a result, DDP instances processing more input batches will hang, as the peer with least input batch will not join those additional allreduce operations. 

Applications can address this by sth like:

```python
for batch in get_batch():
    x = torch.tensor(int(has_next())
    op = all_reduce(x, async_op=True)
    ddp(batch).sum().backward()
    opt.step()
    op.wait()
    if x.item() > 0:
        break
```

As the allrecuce is async and can overlap with the forward+backward+optimizer computation, the extra overhead should be fine. The question is whether we should implement this as an helper API in DDP, or can we leave it out as the application-side solution is simple enough?

@pritamdamania87 mentioned that this solution might not be sufficient for applications that cannot support `has_next()` API. In that case, we would need to support sth like below:

```python
for data in iterator:
  loss = ddp_model(data)
  loss.backward()
  optim.step()

# This writes EOF to the store, and abort allreduce in other DDP instances.
ddp_model.mark_end_of_data()
```

This approach is more versatile, but needs to expose `ProcessGroup` internal `Store` to DDP and also need to implement `abort` for all `ProcessGroup` backends.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar

**Comment by mrshenli** [comment]
We might be able to implement the API suggested by @pritamdamania87's without exposing `Store` or causing (noticeable) perf regression.

* In every DDP forward call, we launch an async allreduce on `torch.tensor(1)` upfront, and record the async_op handle as a DDP member field.
* At the end of ddp forward, wait on the `async_op`. 
    * If the result if `== world_size`, proceed
    * If the result is `< world_size`, then some peer DDP instance has depleted its input. So, all other DDP instances should not proceed with backward either. Set `self.is_end_of_data = True`, and exit loop (or return a dummy output/loss, depends on what API we would like to build).
* In `ddp_model.mark_end_of_data()`, we first check `is_end_of_data`. 
    * If `True`, no op. 
    * If `False`, then it means this DDP instance is the first one exiting the loop. Launch another allreduce on `torch.tensor(0)`. This allreduce will match with all other peer DDP instances' allreduce before forward call, and will lead to the entire gang to exit the loop.

This can be done completed outside of DDP has a helper function, and I would vote for doing it as a helper function instead of as DDP member functions/args. E.g.,

```python
for data in iterator:
  dist.not_end()
  loss = ddp_model(data)
  if not dist.is_gang_end():
    loss.backward()
    optim.step()

dist.end_gang()
```

As the allreduce can overlap with the forward pass, the extra overhead should be negligible.

**Comment by pritamdamania87** [comment]
> As the allreduce can overlap with the forward pass, the extra overhead should be negligible.

In some cases the allreduce might be significantly more expensive than running the forward pass for a simple model. The reason being that allreduce involves network communication compared to CPU/GPU ops locally for the forward pass of a simple model. In such cases, we might actually see some significant performance degradation.

**Comment by pritamdamania87** [comment]
``` 
for data in iterator:
  loss = ddp_model.forward()
  loss.backward()
  optim.step()
ddp_model.mark_end_of_data()
```

>Assume we have two DDP instances (p0, p1), and the first DDP instance finishes earlier. In every iteration, DDP first launches the meta allreduce, and then launches all subsequent bucket allreduces. (we can even create dedicated process group instances for meta allreduce)
Then p0 will hit the mark_end_of_data() call while p1 is still in the loop. So:
p0: launch the meta allreduce in mark_end_of_data() p1: launch meta allreduce and all buckets allreduce in the loop.
the two meta allreduce would match, so that p1 can then abort the bucket allreduces.
(The point here is since we allready have allreduce to collect global information, we might not need to using another form, Store, to do that.)

Regarding this approach, one of the concerns was whether we can abort MPI ops. Looks like there is an API for this purpose: https://www.mpich.org/static/docs/v3.2.x/www3/MPI_Abort.html (would need some testing to see if it behaves as expected though).

**Comment by mrshenli** [comment]
> In some cases the allreduce might be significantly more expensive than running the forward pass for a simple model. The reason being that allreduce involves network communication compared to CPU/GPU ops locally for the forward pass of a simple model. In such cases, we might actually see some significant performance degradation.

Good point! That's true.

> Regarding this approach, one of the concerns was whether we can abort MPI ops. Looks like there is an API for this purpose: https://www.mpich.org/static/docs/v3.2.x/www3/MPI_Abort.html (would need some testing to see if it behaves as expected though).

Thanks for the link! Another concern is where should DDP get the store from. We might want to avoid adding it to the DDP constructor as an additional arg as there are already many of them. Should we let ProcessGroup expose the store instance?

Regarding the API, if we provide `mark_end_of_data()`, it looks like we also need another API to unmark it, e.g., `resume_training()`, in case the application want to reuse the DDP instance for the next epoch. And this would also mean DDP needs to blocking wait for a store read in every iteration? This might cause some regression, but probably is OK as the delay should be short.

**Comment by pritamdamania87** [comment]
> Thanks for the link! Another concern is where should DDP get the store from. We might want to avoid adding it to the DDP constructor as an additional arg as there are already many of them. Should we let ProcessGroup expose the store instance?

Actually I think this approach was without the store, it involves adding a "meta" allreduce as the first allreduce op in the backward pass. It's similar to the approach you proposed above, but we do the allreduce in the backward pass instead of forward (because the backward pass anyways needs to run a bunch of allreduce ops). The only difference is that we need the ability to abort the allreduce ops that have already started.

> Regarding the API, if we provide mark_end_of_data(), it looks like we also need another API to unmark it, e.g., resume_training(), in case the application want to reuse the DDP instance for the next epoch. And this would also mean DDP needs to blocking wait for a store read in every iteration? This might cause some regression, but probably is OK as the delay should be short.

Maybe the APIs should be `mark_end_of_epoch` and `begin_epoch`? 

> And this would also mean DDP needs to blocking wait for a store read in every iteration? This might cause some regression, but probably is OK as the delay should be short.

I didn't mean to use the store, but a "meta" allreduce as part of the backward as you had suggested earlier.

**Comment by rohan-varma** [comment]
Was wondering if we have consensus on how to go about implementing this? If we're going with the `mark_end_of_data`/`mark_end_of_epoch` approach, do we know the process in which the other DDP processes (which presumably are stuck in loss.backward() and have an arbitrary amount of batches left to process) will exit? If we simply abort the allreduce, how will the application know to not run the optimizer.step() and not to continue with processing the rest of the data?

**Comment by pritamdamania87** [comment]
Adding some more information here, looks like horovod built a join op to resolve this: https://github.com/horovod/horovod/issues/832.

**Comment by soulslicer** [comment]
I wanted to post this thread to show how it is solved for now by @mrshenli :

https://discuss.pytorch.org/t/multiprocessing-barrier-blocks-all-processes/80345/27?u=soulslicer

**Comment by rohan-varma** [comment]
Done by (#42577)
